### PR TITLE
[JENKINS-43433] Allow SCMSource implementations to expose merge and origin of change request heads

### PR DESCRIPTION
**Pull request by stephenc**
See [JENKINS-43433](https://issues.jenkins-ci.org/browse/JENKINS-43433)

Todo:

- [x] Update Mock SCM provider to allow for this
- [x] Sanity check that API changes work with GitHub and Bitbucket without causing rebuild storms

### CODE REVIEW

**Comment by stephenc** [comment]
@reviewbybees

File: src/main/java/jenkins/scm/api/mixin/ChangeRequestSCMRevision.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+     */
+    private final SCMRevision target;
+
+    /**
+     * Constructor.
+     *
+     * @param head   the {@link SCMHead} that the {@link SCMRevision} belongs to.
+     * @param target the {@link SCMRevision} of the {@link ChangeRequestSCMHead#getTarget()}.
+     */
+    protected ChangeRequestSCMRevision(@NonNull H head, @NonNull SCMRevision target, boolean merge) {
```
Comment (rsandell) [inline src/main/java/jenkins/scm/api/mixin/ChangeRequestSCMRevision.java]: `merge` is unused

**Comment by stephenc** [comment]
@rsandell I'm more thinking are the names and approach in general right.

This is being slightly driven by the way the GitHub Branch Source has implemented things... which could be the tail wagging the dog (at least for the mergeable mixin)

**Comment by stephenc** [comment]
@rsandell we should be good to go on this once we get a re-review

File: src/main/java/jenkins/scm/api/SCMHeadOrigin.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+     * A named fork. The name should be simple not an URL.
+     */
+    public static class Fork extends SCMHeadOrigin {
+        /**
+         * Standardize serialization.
+         */
+        private static final long serialVersionUID = 1L;
+
+        /**
+         * The name of the fork. This should
```
Comment (jglick) [inline src/main/java/jenkins/scm/api/SCMHeadOrigin.java]: Should what?

File: src/main/java/jenkins/scm/api/mixin/ChangeRequestCheckoutStrategy.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+     * (perhaps with de-duplication).
+     * <p>
+     * In the event that the rebase operation cannot be completed then the checkout operation will fail.
+     * </p>
+     * <p>
+     * It is assumed that the algorithm for rebasing is deterministic given the revision of the change request and
+     * the revision of the {@link ChangeRequestSCMHead#getTarget()}
+     * </p>
+     */
+    REBASE,
```
Comment (jglick) [inline src/main/java/jenkins/scm/api/mixin/ChangeRequestCheckoutStrategy.java]: Why does this need to be defined in the API? IOW why would a client need to differentiation this from `MERGE`? Obviously an implementation could perform any of a variety of merge-like strategies (squashes, for example).

**Pull request merged**